### PR TITLE
Add shift-tab keyboard support for dialogs (modal & Offcanvas components)

### LINKED ISSUE #28481: v5: Modal focus changes
Skipping some previous v4-based PRs given our move to regular JS has already landed in `master`.

- #25277: Modal tabbing and shift+tabbing behavior
- #27646: Allow shift-tab focusing for modals
- #25651: Modal input autofocus attr & keyboard nav
- #21661: shift+tab without focused element within modal
- #27241: SHIFT + TAB should cycle backwards

/cc @Johann-S

### PR DESCRIPTION
**Pull request by RyanBerliner**
Fixes #28481

- Share focus trapping behavior between modal and Offcanvas components
- Allow shift-tab keyboard navigation to wrap around backwards (go to the last focusable element)

There are a couple issues (primarily #33804) that aim to fix inconsistent screen reader navigation or announcements. The work I've done here only concerns tab navigation.

**TODO**:

- [x] Test and revise the `focusableChildren` selector
- [x] Fix `isVisible` utility function (tracking #33914)
- [x] **Fix BrowserStack tests**

[Preview Modal](https://deploy-preview-33865--twbs-bootstrap.netlify.app/docs/5.0/components/modal/)
[Preview Offcanvas](https://deploy-preview-33865--twbs-bootstrap.netlify.app/docs/5.0/components/offcanvas/)

### CODE REVIEW

File: js/src/offcanvas.js
Comment (GeoSot) [inline js/src/offcanvas.js:127]: can this got on line 117 after activation? (or maybe on activation the focusTrap to focus on trapElement?)

File: js/src/util/focustrap.js
Comment (GeoSot) [inline js/src/util/focustrap.js:38]:
```suggestion
    if (!this._config.trapElement || this._isActive) {
```

File: js/src/util/focustrap.js
Comment (GeoSot) [inline js/src/util/focustrap.js:44]: I would love it to see the functionality in another method and simplified

File: js/src/dom/selector-engine.js
Comment (RyanBerliner) [reply js/src/dom/selector-engine.js:88]: Nice! I just did a quick manual test and have found that the implementation of `isVisible` is not thorough enough so it won't work for this use case without fixing it. [Here is a demo of it not working.](https://jsfiddle.net/q4s7Lcr1/) Would you prefer I open a separate issue or deal with it here?

File: js/src/offcanvas.js
Comment (RyanBerliner) [reply js/src/offcanvas.js:127]: I've done it this way to match the initial focus behavior of the modal component. There, it focuses the element when the show transition completes. If I were to move this to line 117, the focus would jump to Offcanvas before it has finished transitioning to shown.

File: js/src/dom/selector-engine.js
Comment (GeoSot) [reply js/src/dom/selector-engine.js:88]: I can confirm the `isVisible` miss-function, and I suppose, is better to fix it on another PR, for easier review

File: js/src/offcanvas.js
Comment (GeoSot) [reply js/src/offcanvas.js:127]: Hmmm, maybe the opposite? (bring `this._focustrap.activate()` on line 126)

File: js/src/offcanvas.js
Comment (GeoSot) [inline js/src/offcanvas.js:125]: now that both focusTrap activations are followed by element focus, you may consider to move it inside activation 👼

Comment (RyanBerliner) [reply js/src/offcanvas.js:125]: Yeah - I had considered that. My hesitation was that it assumes that the `trapElement` itself actually can be focused (tabindex > -2). So, there would need to be additional checks done in the focus trap class before trying to focus itself. Your call, I'm cool with it either way.

Comment (GeoSot) [reply js/src/offcanvas.js:125]: As both components doesn't check something like this, I think you can just add an option `autofocus|focusOnInit` that will be true by default and make it act accordingly

**Comment by GeoSot** [comment]
@RyanBerliner can you rebase your MR please?

**Comment by RyanBerliner** [comment]
@GeoSot I've rebased. It looks like #33327 introduced a dependency of `selector-engine.js` for `util/index.js`. Combine this with my previous work on this PR and it's throwing circular dependency warning between the util index and selector engine files. I'm open to any number of resolutions, though IMO it seems that the `getElement()` in the util index file might be best suited living in the selector engine file? Let me know what you'd prefer to see done.

**Comment by GeoSot** [comment]
>  though IMO it seems that the getElement() in the util index file might be best suited living in the selector engine file? 

Yes, seems right as it has a foreign dependency. Can you handle it (I suppose is better on a separate MR)?

**/cc** @rohit2sharma95 , @alpadev

**Comment by RyanBerliner** [comment]
I hadn't realized how dependent other utility functions were on `getElement()` and `isElement()`... so moving either of those to selector engine would likely turn into some larger, and questionable refactoring. To avoid this I've decided instead to move `focusableChildren()` to the utility file. I think this can be justified since `focusableChildren()` does have some more opinionated logic in it that isn't as straight forward as the other selector engine functions.

**Comment by GeoSot** [comment]
Check the preview link. It misses some things in modal:l89

**Comment by GeoSot** [comment]
Seems good to my eyes.

**CC\** @twbs/js-review

**Review by GeoSot — DISMISSED** [review]

**Comment by GeoSot** [comment]
@RyanBerliner  , can we update this MR (and resolve the suggestion)?

File: js/src/util/focustrap.js
Comment (RyanBerliner) [reply js/src/util/focustrap.js:41]: Yup!

**Comment by XhmikosR** [comment]
@patrickhlauke: LGTY?

**Comment by XhmikosR** [comment]
FYI This doesn't pass on BrowserStack + there are some warnings: https://github.com/twbs/bootstrap/runs/3125887723

We should get those sorted before landing this :)

**Comment by RyanBerliner** [comment]
@XhmikosR the warnings seem simple enough, I can handle them. The failed tests on the other hand seem to be related to differences in how the `<area>` tag is selected. This came up in #33960, and we decided to just drop special handling of this element... I'm wondering if you all think it may be most practical to drop it from the selector here as well since it's used so little. Otherwise, happy to troubleshoot... just want to make sure everyone believes it's worth it since it's a funky element. I'm leaning towards dropping it.

**Comment by GeoSot** [comment]
I will support the 'drop' case, for one more time. However try to leave a comment about it, for future reference

**Review by patrickhlauke — APPROVED** [review]
Testing the behaviour, this now seems to behave correctly. Not looked at the code side, but from a functional point of view, this looks good to me

**Comment by RyanBerliner** [comment]
@XhmikosR I'm not sure I have the permission to re-run browserstack tests, however I believe everything should be resolved. If there are still warnings or errors that you'd like me to revisit lmk.

**Review by GeoSot — APPROVED** [review]

**Comment by XhmikosR** [comment]
All good @RyanBerliner thanks!

Also, thanks to @GeoSot everything seems to be good to go.

**Pull request merged**
